crates/dbsp/src/operator/communication/exchange.rs

```diff
@@ -31,7 +60,188 @@ use std::{
 // using the id as a key.  If there already is an `Exchange` with this id in
 // the store, created by another worker, a reference to that `Exchange` will
 // be used instead.
-circuit_cache_key!(local ExchangeId<T>(usize => Arc<Exchange<T>>));
+circuit_cache_key!(local ExchangeCacheId<T>(ExchangeId => Arc<Exchange<T>>));
+
+#[tarpc::service]
+trait ExchangeService {
+    async fn exchange(data: Vec<u8>, exchange_id: usize, sender: usize, receiver: usize);
+}
+
+type ExchangeId = usize;
+
+// Maps from an `exchange_id` to the `Inner` that implements the exchange.
+type ExchangeDirectory = Arc<RwLock<HashMap<ExchangeId, Arc<Inner>>>>;
+
+#[derive(Clone)]
+struct ExchangeServer(ExchangeDirectory);
+
+#[tarpc::server]
+impl ExchangeService for ExchangeServer {
+    async fn exchange(
+        self,
+        _: context::Context,
+        data: Vec<u8>,
+        exchange_id: ExchangeId,
+        sender: usize,
+        receiver: usize,
+    ) {
+        let inner = self.0.read().unwrap().get(&exchange_id).unwrap().clone();
+        inner.received(data, sender, receiver).await;
+    }
+}
+
+struct Clients(Vec<(Range<usize>, ExchangeServiceClient)>);
+
+impl Clients {
+    fn new(runtime: &Runtime, local_client: &ExchangeServiceClient) -> Clients {
+        Self(match runtime.layout() {
+            Layout::Solo { n_workers } => vec![(0..*n_workers, local_client.clone())],
+            Layout::Multihost {
+                hosts,
+                local_host_idx,
+            } => hosts
+                .iter()
+                .enumerate()
+                .map(|(idx, host)| {
+                    let client = if idx != *local_host_idx {
+                        let mut transport = connect(host.address, Bincode::default);
+                        transport.config_mut().max_frame_length(usize::MAX);
+                        // XXX Blocking here could cause deadlock?  We should connect
+                        // asynchronously.
+                        sleep(std::time::Duration::from_millis(1000));
+                        println!("connecting to {}", host.address);
+                        let transport = TokioHandle::current().block_on(transport).unwrap();
+                        println!("connected to {}", host.address);
+                        ExchangeServiceClient::new(client::Config::default(), transport).spawn()
+                    } else {
+                        local_client.clone()
+                    };
+                    (host.workers.clone(), client)
+                })
+                .collect(),
+        })
+    }
+}
+
+impl Index<usize> for Clients {
+    type Output = ExchangeServiceClient;
+    fn index(&self, worker: usize) -> &ExchangeServiceClient {
+        for (workers, host) in self.0.iter() {
+            if workers.contains(&worker) {
+                return host;
+            }
+        }
+        unreachable!();
+    }
+}
+
+struct Inner {
```

Comment:
I am not fond of such general non-descriptive names. Perhaps this is the rusty way to do things?
Without an IDE such names are hard to use, since they don't tell you anything about their purpose.

Reply:
I would call this InnerExchange, or something similar.

Reply:
OK, I renamed it to InnerExchange.